cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go

```diff
@@ -0,0 +1,255 @@
+	Cfg() *kubeadmapi.JoinConfiguration
+	ClientSetFromFile(string) (*clientset.Clientset, error)
+	InitCfg() (*kubeadmapi.InitConfiguration, error)
+}
+
+func getPhaseDescription(component string) string {
+	return fmt.Sprintf("Generates the %s static Pod manifest", component)
+}
+
+func getControlPlaneJoinPhaseFlags() []string {
+	return []string{
+		options.APIServerAdvertiseAddress,
+		options.APIServerBindPort,
+		options.CfgPath,
+		options.ControlPlane,
+		options.NodeName,
+		options.TokenDiscovery,
+		options.TokenDiscoveryCAHash,
+		options.TokenDiscoverySkipCAHash,
+	}
+}
+
+// NewControlPlaneJoinPhase creates a kubeadm workflow phase that implements marking the new node as master and update the cluster status
+// with information about current node
+func NewControlPlaneJoinPhase() workflow.Phase {
+	return workflow.Phase{
+		Name:  "control-plane-join",
+		Short: "Prepares the machine to join a control plane.",
+		Long:  cmdutil.MacroCommandLongDescription,
+		Phases: []workflow.Phase{
+			{
+				Name:           "all",
+				Short:          "Generates all static Pod manifest files",
+				InheritFlags:   getControlPlaneJoinPhaseFlags(),
+				RunAllSiblings: true,
+			},
+			newControlPlaneJoinSubphase(kubeadmconstants.KubeAPIServer),
+			newControlPlaneJoinSubphase(kubeadmconstants.KubeControllerManager),
+			newControlPlaneJoinSubphase(kubeadmconstants.KubeScheduler),
+			newEtcdLocalSubphase(),
+			newUploadConfigSubphase(),
+			newMarkControlPlaneSubphases(),
+		},
+		InheritFlags: getControlPlaneJoinPhaseFlags(),
+	}
+}
+
+func newControlPlaneJoinSubphase(component string) workflow.Phase {
+	return workflow.Phase{
+		Name:         controlPlanePhaseProperties[component].name,
+		Short:        controlPlanePhaseProperties[component].short,
+		Run:          runControlPlaneJoinSubphase(component),
+		InheritFlags: getControlPlaneJoinPhaseFlags(),
+	}
+}
+
+func newEtcdLocalSubphase() workflow.Phase {
+	return workflow.Phase{
+		Name:         "local",
+		Short:        "Generates the static Pod manifest file for a local, single-node local etcd instance.",
+		Run:          runEtcdLocalPhase,
+		InheritFlags: getControlPlaneJoinPhaseFlags(),
+	}
+}
+
+func newUploadConfigSubphase() workflow.Phase {
+	return workflow.Phase{
+		Name:         "upload-config",
+		Short:        "uploading currently used configuration to the cluster",
+		Run:          runUploadConfigPhaseLocal,
+		InheritFlags: getControlPlaneJoinPhaseFlags(),
+	}
+}
+
+func newMarkControlPlaneSubphases() workflow.Phase {
+	return workflow.Phase{
+		Name:         "mark-control-plane",
+		Short:        "marking the control-plane with right label",
+		Run:          runMarkControlPlanePhaseLocal,
+		InheritFlags: getControlPlaneJoinPhaseFlags(),
+	}
+}
+
+func runEtcdLocalPhase(c workflow.RunData) error {
+	kubeConfigFile := filepath.Join(kubeadmconstants.KubernetesDir, kubeadmconstants.AdminKubeConfigFileName)
+
+	client, err := kubeconfigutil.ClientSetFromFile(kubeConfigFile)
```

Comment:
once you have `data`, you should use this method instead:
https://github.com/kubernetes/kubernetes/blob/master/cmd/kubeadm/app/cmd/join.go#L394

Reply:
done